### PR TITLE
[Identity] Adding Service Fabric and Azure Arc Managed Identity Support

### LINKED ISSUE #13693: Azure.Identity Support for Managed Identity for Azure Arc Applications
Managed identities for Azure resources provide Azure services with an automatically managed identity in Azure Active Directory. Using a managed identity, you can authenticate to any service that supports Azure AD authentication without having credentials in your code. 

https://azure.microsoft.com/en-us/services/azure-arc/

https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/services-support-managed-identities#azure-arc-enabled-kubernetes

### LINKED ISSUE #13695: Azure.Identity Support for Managed Identity for Service Fabric Applications
A common challenge when building cloud applications is how to securely manage the credentials in your code for authenticating to various services without saving them locally on a developer workstation or in source control. Managed identities for Azure solve this problem for all your resources in Azure Active Directory (Azure AD) by providing them with automatically managed identities within Azure AD. You can use a service's identity to authenticate to any service that supports Azure AD authentication, including Key Vault, without any credentials stored in your code.

https://docs.microsoft.com/en-us/azure/service-fabric/concepts-managed-identity

https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/services-support-managed-identities#azure-service-fabric

### PR DESCRIPTION
**Pull request by schaabs**
- Refactor internal interface `IManagedIdentitySource` to base class `ManagedIdentitySource` to support managed identity implementations overriding request behavior.
- Add support for Azure Arc managed identity
- Add support for Service fabric managed identity

Fixes #13695, Fixes #13693

### CODE REVIEW

File: sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+namespace Azure.Identity
+{
+    internal class AzureArcManagedIdentitySource : ManagedIdentitySource
+    {
+
+        private const string IdentityEndpointInvalidUriError = "The environment variable IDENTITY_ENDPOINT contains an invalid Uri.";
+        private const string NoChallengeErrorMessage = "Did not receive expected WWW-Authenticate header in the response from Azure Arc Managed Identity Endpoint.";
+        private const string InvalidChallangeErrorMessage = "The WWW-Authenticate header in the response from Azure Arc Managed Identity Endpoint did not match the expected format.";
+        private const string ArchApiVersion = "2019-11-01";
```
Comment (chlowell) [inline sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs:19]:
```suggestion
        private const string ArcApiVersion = "2019-11-01";
```

File: sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        private AzureArcManagedIdentitySource(CredentialPipeline pipeline, Uri endpoint, string clientId) : base(pipeline)
+        {
+            _endpoint = endpoint;
+            _clientId = clientId;
+        }
+
+        protected override Request CreateRequest(string[] scopes)
+        {
+            // covert the scopes to a resource string
+            string resource = ScopeUtilities.ScopesToResource(scopes);
```
Comment (chlowell) [inline sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs:57]: I recall hearing the endpoint accepts v2 scopes; can someone with a test environment confirm (@mccoyp @g2vinay @catalinaperalta)?

File: sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                }
+
+                var splitChallenge = challenge.Split(new char[] { '=' }, StringSplitOptions.RemoveEmptyEntries);
+
+                if (splitChallenge.Length != 2)
+                {
+                    throw new AuthenticationFailedException(InvalidChallangeErrorMessage);
+                }
+
+                var authHeaderValue = "Basic " + File.ReadAllText(splitChallenge[1]);
```
Comment (chlowell) [inline sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs:91]: Do we get `AuthenticationFailedException` when this fails?

File: sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs (hunk 2 above)
Comment (mccoyp) [reply sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs:57]: I can confirm that v2 scopes work in my environment

File: sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+        protected override Request CreateRequest(string[] scopes)
+        {
+            // covert the scopes to a resource string
+            string resource = ScopeUtilities.ScopesToResource(scopes);
+
+            Request request = Pipeline.HttpPipeline.CreateRequest();
+
+            request.Method = RequestMethod.Get;
+            request.Headers.Add("X-IDENTITY-HEADER", _identityHeaderValue);
```
Comment (mccoyp) [inline sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs:69]: Service Fabric uses a "Secret" header: [docs](https://docs.microsoft.com/en-us/azure/service-fabric/how-to-managed-identity-service-fabric-app-code#acquiring-an-access-token-using-rest-api)
```suggestion
            request.Headers.Add("Secret", _identityHeaderValue);
```

File: sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            request.Headers.Add("Metadata", "true");
+
+            request.Uri.Reset(_endpoint);
+            request.Uri.AppendQuery("api-version", ArchApiVersion);
+
+            request.Uri.AppendQuery("resource", resource);
+
+            if (!string.IsNullOrEmpty(_clientId))
+            {
+                request.Uri.AppendQuery("client_id", _clientId);
```
Comment (g2vinay) [inline sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs:69]: Azure Arc doesn't support user assigned identity.
The req with client id might fail.

File: sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+        private static bool ValidateMsiServerCertificate(HttpRequestMessage message, X509Certificate2 cert, X509Chain certChain, SslPolicyErrors policyErrors)
+        {
+            // Do any additional validation here
+            if (policyErrors == SslPolicyErrors.None)
+            {
+                return true;
+            }
+
+            return 0 == string.Compare(cert.GetCertHashString(), EnvironmentVariables.IdentityServerThumbprint, StringComparison.OrdinalIgnoreCase);
```
Comment (g2vinay) [inline sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs:90]: Does cert.GetCertHashString() return the well formatted thumbprint "XX:XX:XX:XX:XX...." or "XXXXXXXX...." ?
If its well formatted then we'll need to remove the ':' separator before comparing.

File: sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs (hunk 2 above)
Comment (g2vinay) [reply sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs:57]: If by V2 scope you mean, sending the scope with /.default, ex: "https://valut.azure.net/.default"
Then the initial request works fine to get the secret key, but the auth req with the Authorization header set, comes back with 400 when scope with /.default is sent in my Arc Vm.

File: sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs (hunk 5 above)
Comment (g2vinay) [reply sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs:69]: On my sample test, it appears that client_id parameter is ignored on server side, so the req won't fail, but best to not send it, since its not used.

File: sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs (hunk 6 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs:90]: No separators just the hex value.

File: sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs (hunk 3 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs:91]: Yes any exceptions resulting from reading the file will be handled by the credential and wrapped in an `AuthenticationFailedException`.

File: sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs (hunk 2 above)
Comment (mccoyp) [reply sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs:57]: I was able to get a deleted secret with something like:
```Python
credential = ManagedIdentityCredential()
token = credential.get_token("https://vault.azure.net/.default")
client = SecretClient("vault_url", credential)
client.get_deleted_secret("TestSecret")
```
Does this work to validate the behavior, or is there another way I should test v2 scopes?

File: sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs (hunk 1 above)
Comment (mccoyp) [reply sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs:19]: Also, we just got confirmation from the service team that we can target the latest version of `2020-06-01`

File: sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs (hunk 2 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/AzureArcManagedIdentitySource.cs:64]: The public API for `ManagedIdentityCredential` does take in scopes, but these are internally converted to resource strings. Testing it out it I confirmed the endpoint only accepts resource strings similarly to all other MI endpoints, @g2vinay  also independently confirmed this.

**Review by laiapat — COMMENTED** [review]
Is it possible to do mock tests for the Service Fabric credential?

**Review by sadasant — APPROVED** [review]
The code looks ok! The logic seems sound. Looks good to me 😄

File: sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+            if (string.IsNullOrEmpty(identityEndpoint) || string.IsNullOrEmpty(identityHeader) || string.IsNullOrEmpty(identityServerThumbprint))
+            {
+                return default;
+            }
+
+            Uri endpointUri;
+            try
+            {
+                endpointUri = new Uri(identityEndpoint);
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs:36]: https://docs.microsoft.com/en-us/dotnet/api/system.uri.trycreate?view=net-5.0 ?

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialArcLiveTests.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+        private class ArcMiRecordedTestMatcher : RecordMatcher
+        {
+            public override RecordEntry FindMatch(RecordEntry request, IList<RecordEntry> entries)
+            {
+                var match = base.FindMatch(request, entries);
+
+                if (match.Response.Headers.TryGetValue("WWW-Authenticate", out string[] challenges) && challenges[0].StartsWith("Basic realm="))
+                {
+                    challenges[0] = challenges[0].Split('=')[0] + "=" + Path.Combine(TestContext.CurrentContext.TestDirectory, "Data", "mock-arc-mi-key.key");
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialArcLiveTests.cs:76]: We generally don't use matcher to modify the record. Why is this required?

File: sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+            return request;
+        }
+
+        private static bool ValidateMsiServerCertificate(HttpRequestMessage message, X509Certificate2 cert, X509Chain certChain, SslPolicyErrors policyErrors)
+        {
+            // Do any additional validation here
+            if (policyErrors == SslPolicyErrors.None)
+            {
+                return true;
```
Comment (chlowell) [inline sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs:97]: So if we trust the cert, we don't care whether it's the one we expect?

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialArcLiveTests.cs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+            if (string.IsNullOrEmpty(TestEnvironment.ArcEnable))
+            {
+                Assert.Ignore();
+            }
+
+            using (ReadOrRestoreManagedIdentityEnvironment())
+            {
+                var vaultUri = new Uri(TestEnvironment.SystemAssignedVault);
+
+                var cred = new ManagedIdentityCredential(options: InstrumentClientOptions(new TokenCredentialOptions()));
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialArcLiveTests.cs:35]: Don't we need to call InstumentClient for  cred here?

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialSFLiveTests.cs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+                var vaultUri = new Uri(TestEnvironment.SystemAssignedVault);
+
+                CredentialPipeline pipeline = CredentialPipeline.GetInstance(InstrumentClientOptions(new TokenCredentialOptions { Transport = ServiceFabricManagedIdentitySource.GetServiceFabricMITransport() }));
+
+                var cred = new ManagedIdentityCredential(new ManagedIdentityClient(new ManagedIdentityClientOptions { Pipeline = pipeline, PreserveTransport = (Mode == RecordedTestMode.Playback) }));
+
+                // Hard code service version or recorded tests will fail: https://github.com/Azure/azure-sdk-for-net/issues/10432
+                var kvoptions = InstrumentClientOptions(new SecretClientOptions(SecretClientOptions.ServiceVersion.V7_0));
+
+                var kvclient = new SecretClient(vaultUri, cred, kvoptions);
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialSFLiveTests.cs:38]: Do we really need to involve the SecretClient in these tests?

**Review by chlowell — APPROVED** [review]

**Review by pakrym — APPROVED** [review]

File: sdk/identity/Azure.Identity/tests/IdentityTestEnvironment.cs [hunk 12]
```diff
@@ -31,5 +42,15 @@ public IdentityTestEnvironment() : base("identity")
         public string ServicePrincipalCertificatePfxPath => GetOptionalVariable("IDENTITY_SP_CERT_PFX") ?? Path.Combine(TestContext.CurrentContext.TestDirectory, "Data", "cert.pfx");
         public string ServicePrincipalCertificatePemPath => GetOptionalVariable("IDENTITY_SP_CERT_PEM") ?? Path.Combine(TestContext.CurrentContext.TestDirectory, "Data", "cert.pem");
         public string ServicePrincipalSniCertificatePath => GetOptionalVariable("IDENTITY_SP_CERT_SNI") ?? Path.Combine(TestContext.CurrentContext.TestDirectory, "Data", "cert.pfx");
+
+        public void RecordOrRestoreManagedIdentityEnvironmentVariables()
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/IdentityTestEnvironment.cs:46]: Not used.

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialSFLiveTests.cs (hunk 11 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialSFLiveTests.cs:38]: Unfortunately yes to ensure we've gotten a token for the right identity. Some managed identity endpoints will unfortunately ignore the specified user assigned client id and give you back a token for the system assigned identity (which sometimes isn't event valid). So we use the Key Vault service as our "token validator".

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialArcLiveTests.cs (hunk 8 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialArcLiveTests.cs:76]: I'm updating the recording in the matcher in this case because the service is responding with a header which contains a path to a file on the local disk. So the idea here is to replace the path the service returned with a static file binplaced when we build the tests. But that path differs depending on the machine your running the tests on, so I can't update it when scrubbing. I originally tried to update the header in the scrubber to be a relative path, but this introduces failures depending on how the tests are run and the working directly.

File: sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs (hunk 9 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs:97]: This validation callback came from the sample provided by the service team, https://docs.microsoft.com/en-us/azure/service-fabric/how-to-managed-identity-service-fabric-app-code#acquiring-an-access-token-using-c. I can reach out to them to see if removing this is makes sense.

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialArcLiveTests.cs (hunk 10 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialArcLiveTests.cs:35]: I've instrumented the `SecretClient` which calls into this credential, both cannot be instrumented since the sync methods on `SecretClient` will call into `GetToken` and not `GetTokenAsync`

File: sdk/identity/Azure.Identity/tests/IdentityTestEnvironment.cs (hunk 12 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/tests/IdentityTestEnvironment.cs]: removed
